src/browser/wisp_network.js

```diff
@@ -0,0 +1,570 @@
+        this.wispws.onerror = () => {
+            register_ws();
+        };
+        this.wispws.onclose = () => {
+            register_ws();
+        };
+    };
+    register_ws();
+    config = config || {};
+    this.bus = bus;
+    this.id = config.id || 0;
+    this.router_mac = new Uint8Array((config.router_mac || "52:54:0:1:2:3").split(":").map(function(x) { return parseInt(x, 16); }));
+    this.router_ip = new Uint8Array((config.router_ip || "192.168.86.1").split(".").map(function(x) { return parseInt(x, 10); }));
+    this.vm_ip = new Uint8Array((config.vm_ip || "192.168.86.100").split(".").map(function(x) { return parseInt(x, 10); }));
+    this.masquerade = config.masquerade === undefined || !!config.masquerade;
+    this.vm_mac = new Uint8Array(6);
+
+    this.tcp_conn = {};
+
+    this.bus.register("net" + this.id + "-mac", function(mac) {
+        this.vm_mac = new Uint8Array(mac.split(":").map(function(x) { return parseInt(x, 16); }));
+    }, this);
+    this.bus.register("net" + this.id + "-send", function(data)
+    {
+        this.send(data);
+    }, this);
+}
+
+WispNetworkAdapter.prototype.destroy = function()
+{
+    if(this.wispws)
+        this.wispws.close();
+};
+
+// https://stackoverflow.com/questions/4460586/javascript-regular-expression-to-check-for-ip-addresses
+function validate_IP_address(ipaddress) {
+    if(/^(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/.test(ipaddress)) {
+    return true;
+    }
+    return false;
+}
+
+// DNS over HTTPS fetch, recursively fetch the A record until the first result is an IPv4
+async function dohdns(q) {
+    const req = await fetch(`${DOH_SERVER}?name=${q.name.join(".")}&type=${q.type}`, {headers: [["accept", "application/dns-json"]]});
+    if(req.status === 200) {
+        const res = await req.json();
+        if(res.Answer) {
+            if(validate_IP_address(res.Answer[0].data)) {
+                return res;
+            } else {
+                return await dohdns({name: res.Answer[0].data.split("."), type: q.type});
+            }
+        }
+        return { // if theres an error, Naively return localhost
+            "Status": 0,
+            "TC": false,
+            "RD": true,
+            "RA": true,
+            "AD": true,
+            "CD": false,
+            "Question": [
+                {
+                    "name": q.name.join("."),
+                    "type": 1
+                }
+            ],
+            "Answer": [
+                {
+                    "name": q.name.join("."),
+                    "type": 1,
+                    "TTL": 600,
+                    "data": "127.0.0.1"
+                }
+            ]
+        };
+    } else {
+        dbg_log("DNS Server returned error code: " + await req.text(), LOG_NET);
+    }
+}
+
+/**
+ * @param {Uint8Array} data
+ */
+WispNetworkAdapter.prototype.send = function(data)
+{
```

Comment:
Its' mainly this logic that I worry should be refactored out.

Reply:
What do you have in mind? In general I think the way the wisp client will handle TCP and UDP too differently for that function to be shared between fetch_network and wisp_network, but I'm open to ideas on how it should be organized

Reply:
it looks similar at first glance but if you look a bit more at packet.tcp.syn you'll see that I modified a bit of what goes into this.tcp_conn\[tuple\] like adding stream_id and send_wisp_frame, those weren't originally a part of the function in the fetch_network I believe

Reply:
I think a good place to start would be to extract the duplicate logic for things like dhcp and ntp into functions in fetch-networking and then just call them in the handler.

Reply:
I can do that, should I make a new file for this?

Reply:
Yeah I was debating on putting all the protocol utility stuff it its own file independent of the "clever" fetch impersonation bits, but ultimately decided against.  If more than one file is going to use them it might make sense though.

Reply:
@copy thoughts before I (or if I should)  split the shared "utility" functions into net-utils.js?

Reply:
> Yeah I was debating on putting all the protocol utility stuff it its own file independent of the "clever" fetch impersonation bits, but ultimately decided against. If more than one file is going to use them it might make sense though.

I'm refactoring both right now and I see FetchNetworkAdapter.prototype.tcp_connect but it seems to be fully unused, tests still run perfectly without it too, should I get rid of it? or are there future plans for it